engine/sql/core/src/main/scala/org/apache/spark/sql/partitioner/QuadTreePartitioner.scala

```diff
@@ -0,0 +1,142 @@
+    } else {
+      origin.mapPartitions { iter =>
+        val mutablePair = new MutablePair[Point, InternalRow]()
+        iter.map(row => mutablePair.update(row._1, row._2.copy()))
+      }
+    }
+
+    val part = new QuadTreePartitioner(est_partition, sample_rate, dimension,
+      transfer_threshold, max_entries_per_node, rdd)
+    val shuffled = new ShuffledRDD[Point, InternalRow, InternalRow](rdd, part)
+    shuffled.setSerializer(new SparkSqlSerializer(new SparkConf(false)))
+    (shuffled, part.mbrBound)
+  }
+}
+
+class QuadTreePartitioner(est_partition: Int,
+                          sample_rate: Double,
+                          dimension: Int,
+                          transfer_threshold: Long,
+                          max_entries_per_node: Int,
+                          rdd: RDD[_ <: Product2[Point, Any]])
+  extends Partitioner {
+  private case class Bounds(min: Array[Double], max: Array[Double])
+
+  var (mbrBound, partitions) = {
+    val (data_bounds, total_size) = {
+      rdd.aggregate[(Bounds, Long)]((null, 0))((bound, data) => {
+        val new_bound = if (bound._1 == null) {
+          new Bounds(data._1.coord, data._1.coord)
+        } else {
+          new Bounds(bound._1.min.zip(data._1.coord).map(x => Math.min(x._1, x._2)),
+            bound._1.max.zip(data._1.coord).map(x => Math.max(x._1, x._2)))
+        }
+        (new_bound, bound._2 + SizeEstimator.estimate(data._1))
+      }, (left, right) => {
+        val new_bound = {
+          if (left._1 == null) right._1
+          else if (right._1 == null) left._1
+          else {
+            new Bounds(left._1.min.zip(right._1.min).map(x => Math.min(x._1, x._2)),
+              left._1.max.zip(right._1.max).map(x => Math.max(x._1, x._2)))
+          }
+        }
+        (new_bound, left._2 + right._2)
+      })
+    } // get the partition bound and the total size of a MBR
+
+    val seed = System.currentTimeMillis()
+    val sampled = if (total_size * sample_rate <= transfer_threshold){
+      rdd.sample(withReplacement = false, sample_rate, seed).map(_._1).collect()
+    }
+    else {
+      rdd.sample(withReplacement = true, transfer_threshold / total_size, seed).map(_._1).collect()
+    }
+
+    def recursiveGroupPoint(entries: Array[Point],
+                            low_bound: Seq[Double], high_bound: Seq[Double]): Array[MBR] = {
+      val center: Seq[Double] = low_bound.zip(high_bound).map(a => (a._1 + a._2) / 2.0)
+      var ans = mutable.ArrayBuffer[MBR]()
+      val grouped = entries.groupBy(p => {
+        var quadrant_number = 0 // use a binary number to represent a quadrant
+        for (cur: Int <- center.indices) {
+          if (p.coord(cur) >= center(cur)) {quadrant_number += 1}
+          quadrant_number <<= 1
+        }
+        if (center.length != 1) quadrant_number >>= 1
+        quadrant_number
+      })
+      for (item <- grouped) {
+        if (item._2.length > max_entries_per_node) { // split node
```

Comment:
@Skyprophet I think I need a parameter for a partition with the same function. If I build a KDTree or QuadTree based on the current partition method, the same parameter is needed. I think we don't need to assign a separate parameter in `SQLConf` for each partition method, so I use the same parameter..

BTW, do you have some idea about the name of the existing parameter to choose partition algorithm? I think I cannot even agree with myself about the name.